File: objc/ios-test/LeadDetailViewController.m
//
//  LeadDetailViewController.m
//  ios-test
//
//  Created by Patricia Marie Cesar on 5/23/16.
//  Copyright © 2016 ServiceSeeking. All rights reserved.
//

#import "LeadDetailViewController.h"

#import "Lead.h"
#import "API+Leads.h"
#import "MBProgressHUD+Loading.h"
#import "NSString+Helper.h"

@interface LeadDetailViewController ()

@property (weak, nonatomic) IBOutlet UILabel *daysLeftLabel;
@property (weak, nonatomic) IBOutlet UILabel *jobNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *isUrgentLabel;
@property (weak, nonatomic) IBOutlet UILabel *jobDescriptionLabel;
@property (weak, nonatomic) IBOutlet UILabel *suburbanNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *userNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *biddingEndDate;
@property (weak, nonatomic) IBOutlet UILabel *isNewLabel;

@end

@implementation LeadDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [MBProgressHUD showLoadingHUDAddedTo:self.navigationController.view labelText:@"Loading" detailLabelText:@"Please wait . . ."];
    
    [[API sharedClient] getLeadWithID:self.leadID completionHandler:^(NSDictionary *responseDictionary) {
        
        Lead *lead = [[Lead alloc] initWithDataDictionary:responseDictionary[kData]];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateInfoWithLead:lead];
            [MBProgressHUD hideAllHUDsForView:self.navigationController.view animated:YES];
        });
    }];
}

#pragma mark - UI

- (void)updateInfoWithLead:(Lead *)lead {
    self.biddingEndDate.text = [NSString stringWithFormat:@"Bidding closes on: %@", [NSDateFormatter localizedStringFromDate:[lead.biddingClosesOn toDate] dateStyle:NSDateFormatterMediumStyle timeStyle:NSDateFormatterShortStyle]];
    self.daysLeftLabel.text = [self numberOfDaysAndHoursLeft:[lead.biddingClosesOn toDate]];
    self.jobNameLabel.text = lead.name;
    self.isUrgentLabel.hidden = !lead.isUrgent;
    self.isNewLabel.hidden = !lead.isNew;
    self.jobDescriptionLabel.text = lead.desc;
    self.suburbanNameLabel.text = lead.suburbName;
    self.userNameLabel.text = lead.userName;
}

#pragma mark - Days and hours remaining

- (NSString *)numberOfDaysAndHoursLeft:(NSDate *)date {
    
    if (date != nil) {
        NSDateComponents *components = [[NSCalendar currentCalendar] components:NSCalendarUnitDay fromDate:[NSDate date] toDate:date options:0];
        NSInteger days = components.day;
        NSInteger hours = days % 24;
        
        NSMutableString *daysRemaining = [[NSMutableString alloc] init];
        
        if (days > 0) {
            [daysRemaining appendFormat:@"%ld day", (long)days];
            
            if (days > 1) {
                [daysRemaining appendString:@"s"];
            }
        }
        if (hours > 0) {
            [daysRemaining appendFormat:@" %ld hour", (long)hours];
            
            if (hours > 1) {
                [daysRemaining appendString:@"s"];
            }
        }
        return daysRemaining.copy;
    }
    return @"";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/ios-test/LeadDetailViewController.swift
//
//  LeadDetailViewController.swift
//  ios-test
//
//  Created by Garri Adrian Nablo on 2/13/16.
//  Copyright © 2016 ServiceSeeking. All rights reserved.
//

import UIKit

class LeadDetailViewController: UIViewController {

    private let viewModel: LeadDetailViewModel!
    
    init(lead: Lead) {
        self.viewModel = LeadDetailViewModel(lead: lead)
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.title = "Lead Detail"
        self.edgesForExtendedLayout = .None
        self.view.backgroundColor = .whiteColor()
        
        let scrollView = UIScrollView()
        scrollView.translatesAutoresizingMaskIntoConstraints = false
        self.view.addSubview(scrollView)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: scrollView, attribute: .Top, relatedBy: .Equal, toItem: self.view, attribute: .Top, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: scrollView, attribute: .Left, relatedBy: .Equal, toItem: self.view, attribute: .Left, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: scrollView, attribute: .Right, relatedBy: .Equal, toItem: self.view, attribute: .Right, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: scrollView, attribute: .Bottom, relatedBy: .Equal, toItem: self.view, attribute: .Bottom, multiplier: 1.0, constant: 0.0)
            ])
        
        let contentView = UIView()
        contentView.translatesAutoresizingMaskIntoConstraints = false
        scrollView.addSubview(contentView)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: contentView, attribute: .Top, relatedBy: .Equal, toItem: scrollView, attribute: .Top, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: contentView, attribute: .Left, relatedBy: .Equal, toItem: scrollView, attribute: .Left, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: contentView, attribute: .Right, relatedBy: .Equal, toItem: scrollView, attribute: .Right, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: contentView, attribute: .Bottom, relatedBy: .Equal, toItem: scrollView, attribute: .Bottom, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: contentView, attribute: .CenterX, relatedBy: .Equal, toItem: scrollView, attribute: .CenterX, multiplier: 1.0, constant: 0.0),
            NSLayoutConstraint(item: contentView, attribute: .CenterY, relatedBy: .Equal, toItem: scrollView, attribute: .CenterY, multiplier: 1.0, constant: 0.0)
            ])
        
        let usernameLabel = UILabel()
        usernameLabel.textAlignment = .Left
        usernameLabel.text = viewModel.lead.username
        usernameLabel.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(usernameLabel)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: usernameLabel, attribute: .Top, relatedBy: .Equal, toItem: self.view, attribute: .Top, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: usernameLabel, attribute: .Left, relatedBy: .Equal, toItem: self.view, attribute: .Left, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: usernameLabel, attribute: .Right, relatedBy: .Equal, toItem: self.view, attribute: .Right, multiplier: 1.0, constant: -20.0)
            ])
        
        let nameLabel = UILabel()
        nameLabel.numberOfLines = 0
        nameLabel.textAlignment = .Left
        nameLabel.text = viewModel.lead.name
        nameLabel.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(nameLabel)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: nameLabel, attribute: .Top, relatedBy: .Equal, toItem: usernameLabel, attribute: .Bottom, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: nameLabel, attribute: .Left, relatedBy: .Equal, toItem: self.view, attribute: .Left, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: nameLabel, attribute: .Right, relatedBy: .Equal, toItem: self.view, attribute: .Right, multiplier: 1.0, constant: -20.0)
            ])
        
        let descriptionLabel = UILabel()
        descriptionLabel.numberOfLines = 0
        descriptionLabel.textAlignment = .Left
        descriptionLabel.text = viewModel.lead.jobDescription
        descriptionLabel.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(descriptionLabel)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: descriptionLabel, attribute: .Top, relatedBy: .Equal, toItem: nameLabel, attribute: .Bottom, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: descriptionLabel, attribute: .Left, relatedBy: .Equal, toItem: self.view, attribute: .Left, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: descriptionLabel, attribute: .Right, relatedBy: .Equal, toItem: self.view, attribute: .Right, multiplier: 1.0, constant: -20.0)
            ])
        
        let suburbLabel = UILabel()
        suburbLabel.textAlignment = .Left
        suburbLabel.text = viewModel.lead.suburb
        suburbLabel.translatesAutoresizingMaskIntoConstraints = false
        contentView.addSubview(suburbLabel)
        NSLayoutConstraint.activateConstraints([
            NSLayoutConstraint(item: suburbLabel, attribute: .Top, relatedBy: .Equal, toItem: descriptionLabel, attribute: .Bottom, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: suburbLabel, attribute: .Left, relatedBy: .Equal, toItem: self.view, attribute: .Left, multiplier: 1.0, constant: 20.0),
            NSLayoutConstraint(item: suburbLabel, attribute: .Right, relatedBy: .Equal, toItem: self.view, attribute: .Right, multiplier: 1.0, constant: -20.0)
            ])
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}
